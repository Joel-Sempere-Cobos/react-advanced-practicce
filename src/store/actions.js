import { areAdvertsLoaded, getAdvertByIdRedux } from './selectors.js';
import {
  ADVERTS_LOADED_FAILURE,
  ADVERTS_LOADED_REQUEST,
  ADVERTS_LOADED_SUCCESS,
  ADVERT_LOADED_FAILURE,
  ADVERT_LOADED_REQUEST,
  ADVERT_LOADED_SUCCESS,
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT,
  UI_RESET_ERROR,
} from './types.js';

////////// AUTH_LOGIN

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});

export const authLoginSuccess = () => ({
  type: AUTH_LOGIN_SUCCESS,
});

export const authLoginFailure = (error) => ({
  type: AUTH_LOGIN_FAILURE,
  payload: error,
  error: true,
});

export const authLogin = (credentials, rememberMe) => {
  return async function (dispatch, getState, { api }) {
    try {
      dispatch(authLoginRequest());
      await api.auth.login(credentials, rememberMe);
      dispatch(authLoginSuccess());
    } catch (error) {
      dispatch(authLoginFailure(error));
      throw error;
    }
  };
};
/////////// AUTH_LOGOUT

export const authLogout = () => ({
  type: AUTH_LOGOUT,
});

/////////// ADVERTS

export const advertsLoadedRequest = () => ({
  type: ADVERTS_LOADED_REQUEST,
});

export const advertsLoadedSuccess = (adverts) => ({
  type: ADVERTS_LOADED_SUCCESS,
  payload: adverts,
});

export const advertsLoadedFailure = (error) => ({
  type: ADVERTS_LOADED_FAILURE,
  payload: error,
  error: true,
});

export const advertsLoad = () => {
  return async function (dispatch, getState, { api }) {
    const areLoaded = areAdvertsLoaded(getState());
    if (areLoaded) return;
    try {
      dispatch(advertsLoadedRequest());
      const adverts = await api.adverts.getAdverts();
      dispatch(advertsLoadedSuccess(adverts));
    } catch (error) {
      dispatch(advertsLoadedFailure(error));
      throw error;
    }
  };
};

/////////// ADVERT

export const advertLoadedRequest = () => ({
  type: ADVERT_LOADED_REQUEST,
});

export const advertLoadedSuccess = (adverts) => ({
  type: ADVERT_LOADED_SUCCESS,
  payload: adverts,
});

export const advertLoadedFailure = (error) => ({
  type: ADVERT_LOADED_FAILURE,
  payload: error,
  error: true,
});

export const advertLoad = (id) => {
  return async function (dispatch, getState, { api }) {
    const isLoaded = getAdvertByIdRedux(id)(getState());
    if (isLoaded) return;
    try {
      dispatch(advertLoadedRequest());
      const advert = await api.adverts.getAdvertById(id);
      dispatch(advertLoadedSuccess(advert));
    } catch (error) {
      dispatch(advertLoadedFailure(error));
      throw error;
    }
  };
};

//////////// UI

export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});
