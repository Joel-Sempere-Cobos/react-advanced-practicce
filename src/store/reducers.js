import { ADVERTS_LOADED, AUTH_LOGIN, AUTH_LOGOUT } from './types.js';

const defaultState = {
  auth: false,
  adverts: [],
};

/* export default function reducer(state = defaultState, action) {
  switch (action.type) {
    case AUTH_LOGIN:
      return { ...state, auth: true };
    case AUTH_LOGOUT:
      return { ...state, auth: false };
    case ADVERTS_LOADED:
      return { ...state, adverts: action.payload };
    default:
      return state;
  }
} */

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN:
      return true;
    case AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
}

export function adverts(state = defaultState.adverts, action) {
  switch (action.type) {
    case ADVERTS_LOADED:
      return action.payload;
    default:
      return state;
  }
}

/* export default function reducer(state = defaultState, action) {
  return {
    auth: auth(state.auth, action),
    adverts: adverts(state.adverts, action),
  };
} */
