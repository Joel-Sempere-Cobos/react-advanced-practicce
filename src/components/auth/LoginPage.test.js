import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { defaultState } from '../../store/reducers.js';
import LoginPage from './LoginPage.js';
import { authLogin } from '../../store/actions.js';

jest.mock('../../store/actions.js');

describe('LoginPage', () => {
  const store = {
    getState: () => defaultState,
    dispatch: () => {},
    subscribe: () => {},
  };

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <BrowserRouter>
          <LoginPage />
        </BrowserRouter>
      </Provider>
    );

  test('snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('should dispatch "authLogin" action', async () => {
    const email = 'test@test.com';
    const password = 'prueba';
    const rememberMe = false;

    renderComponent();
    const emailInput = screen.getByLabelText(/Email/);
    const passwordInput = screen.getByLabelText(/Contrase√±a/);
    const submitButton = screen.getByRole('button');

    expect(submitButton).toBeDisabled();

    fireEvent.change(emailInput, { target: { value: email } });
    fireEvent.change(passwordInput, { target: { value: password } });

    expect(submitButton).toBeEnabled();

    fireEvent.click(submitButton);
    expect(authLogin).toHaveBeenCalled();
    expect(authLogin).toHaveBeenCalledWith({ email, password }, rememberMe);
  });
});
