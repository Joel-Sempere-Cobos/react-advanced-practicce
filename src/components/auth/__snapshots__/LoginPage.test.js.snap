// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="form-page-container"
  >
    <h1>
      Bienvenido@ a Wakapop
    </h1>
    <form
      class="form"
    >
      <div
        class="email-form"
      >
        <label
          for="email"
        >
          Email
        </label>
        <input
          id="email"
          name="email"
          type="email"
          value=""
        />
      </div>
      <div
        class="password-form"
      >
        <label
          for="password"
        >
          Contraseña
        </label>
        <input
          id="password"
          name="password"
          type="password"
          value=""
        />
      </div>
      <div
        class="rememberMe-form"
      >
        <label
          for="rememberMe"
        >
          Recuérdame
        </label>
        <input
          id="rememberMe"
          name="rememberMe"
          type="checkbox"
        />
      </div>
      <div
        class="submit-form"
      >
        <button
          disabled=""
          type="submit"
        >
          Login
        </button>
      </div>
    </form>
  </div>
</div>
`;
